elastic/kibana
# [Lens][TSVB][Agg based] TSDB warning handling support for Lens, Agg based and TSVB

Fixes https://github.com/elastic/kibana/issues/132045

Bonus (because it's very easy to do with the new warning handling): Do not show warnings anymore for suggestions. In almost all cases they are just repeating the warnings from the main chart, rendering the config panel unusable and not adding any meaningful information.

## How to test

* Check out https://github.com/elastic/elasticsearch/tree/kibana-downsampling-branch next to your kibana working copy
* Start ES with yarn es source instead of snapshot (this will build the given branch)
* Clone https://github.com/csoulios/downsampling-demo
* Run the commands in [downsample-index.es](http://downsample-index.es/) (indexing the actual data is too much for dev console, so you have to do this via curl curl -XPOST http://elastic:changeme@localhost:9200/sample-01/_bulk\?pretty -H 'Content-Type: application/json' --data-binary @sample-01-raw.json)
* After running all commands, there should be sample-01-rollup which has the rolled up data
* Create a data view `sample-01` for testing TSDB without rollups and a data view `sample-01,sample-01-rollup` for testing TSDB with rollups

## Behavior

In the Lens editor, rollup errors are shown as warnings in the existing warning menu. Lens warnings get a nice message explaining what's going on.
<img width="1246" alt="Screenshot 2022-09-01 at 12 44 22" src="https://user-images.githubusercontent.com/1508364/187897311-03bd5ff8-f8d9-40a8-82b3-521890a44be9.png">

On dashboard, a warning indicator is rendered in the bottom right of the vis
<img width="1294" alt="Screenshot 2022-09-01 at 12 44 28" src="https://user-images.githubusercontent.com/1508364/187897413-84c5a1b8-5576-4890-972a-d58aa6c2fd30.png">


For TSVB and agg based, the warning indicator is rendered both in the editor and on the dashboard. The error message coming from Elasticsearch is not rewritten in the context of the chart.
<img width="1905" alt="Screenshot 2022-09-01 at 12 45 10" src="https://user-images.githubusercontent.com/1508364/187897486-6da62880-8607-4bbc-914b-68f25d811677.png">
<img width="961" alt="Screenshot 2022-09-01 at 12 45 00" src="https://user-images.githubusercontent.com/1508364/187897489-06d326c3-2ea0-4f7f-bfd2-41809424a893.png">

If there are other shard failures than rollup related ones, the existing warning toast is still shown (in this case the shard failures modal is also including the failures originating from rollup if both are happening at the same time)
<img width="805" alt="Screenshot 2022-09-01 at 12 48 45" src="https://user-images.githubusercontent.com/1508364/187897579-29f1ee3b-8ab7-4c9b-8630-0c2e51a4ac23.png">
<img width="411" alt="Screenshot 2022-09-01 at 12 48 38" src="https://user-images.githubusercontent.com/1508364/187897582-60b94e2b-ed85-4b30-8ef6-5ae73d4123f8.png">

To cause the other shard failures, extend the data view to include another index name besides `sample-01` and `sample-01-rollup`, then index a document which auto-sets a used number field to text, e.g. like this:
```
POST other-trickery-1/_doc
{
  "kubernetes.container.memory.available.bytes": "abc",
  "@timestamp": "2022-06-18"
}
```

This will cause a generic illegal argument exception which should not be handled locally.


## Implementation details

Previously the warning toasts would be shown by the esaggs function implementation automatically. This PR changes this in the following way:
* Disable the automatic handling by passing `disableShardWarnings` to the search context
* In the embeddable (visualize and Lens), call `showWarnings` with the adapters and pass a callback which filters out rollup-related warnings

Due to this, the way shard failure warnings can be suppressed for TSVB needs to be changed:
* Do not call `handleWarnings` from within the TSVB request handler anymore - the visualize embeddable does that already
* Add a `suppressWarnings` function to the vis type definition which is checked in the embeddable whether the warning should be shown or not
